### ISSUE TITLE
Empty set from `set()` assignable to `MySet` but not to `MySet | None`

### ISSUE THREAD

**Issue by BHSPitMonkey** [astral-sh/ty#1603]
### Summary

Playground: https://play.ty.dev/b6d2c6ad-1986-425e-9619-12434e33a223

```python
type StyleSet = set[
    Literal[
        "bold",
        "italic",
        "strike",
    ]
]

s: StyleSet = set()
# Works as expected

s: StyleSet | None = set()
# Object of type `set[Unknown]` is not assignable to `StyleSet | None` (invalid-assignment)
```

The custom subscripted `set` subtype above has no trouble accepting an empty `set()`, but `ty` produces an `invalid-assignment` when assigned to `StyleSet | None` (despite this union being wider)

### Version

eb7c098d6

**Comment by carljm** [comment]
@ibraheemdev Haven't looked into this in depth, but it certainly looks like a bidi issue -- even if we infer `set[Unknown]` for the RHS (and it seems like we should be able to do better than that with the annotation?) it should definitely be assignable to `StyleSet | None`.

**Comment by ibraheemdev** [comment]
This should be resolved by https://github.com/astral-sh/ruff/pull/21574. We are correctly using the type context, but promoting to `set[str]` because we don't see through the union. The confusing diagnostic is because of reinference — we should probably to prioritize https://github.com/astral-sh/ruff/pull/21267 to avoid cases like this.
